media/js/base/fxa-utm-referral.js

```diff
@@ -0,0 +1,100 @@
+     * https://mozilla.github.io/application-services/docs/accounts/metrics.html#descriptions-of-metrics-related-query-parameters
+     * @returns {Object} if one or more valit utms exist, else {null}.
+     */
+    UtmUrl.getAttributionData = function () {
+        var params = new window._SearchParams().utmParams();
+        var allowedChars = /^[\w/.%-]+$/;
+        var finalParams = {};
+
+        if (params.hasOwnProperty('utm_source') && (allowedChars).test(params['utm_source'])) {
+            finalParams['utm_source'] = decodeURIComponent(params['utm_source']);
+        }
+
+        if (params.hasOwnProperty('utm_campaign') && (allowedChars).test(params['utm_campaign'])) {
+            finalParams['utm_campaign'] = decodeURIComponent(params['utm_campaign']);
+        }
+
+        if (params.hasOwnProperty('utm_content') && (allowedChars).test(params['utm_content'])) {
+            finalParams['utm_content'] = decodeURIComponent(params['utm_content']);
+        }
+
+        if (params.hasOwnProperty('utm_term') && (allowedChars).test(params['utm_term'])) {
+            finalParams['utm_term'] = decodeURIComponent(params['utm_term']);
+        }
+
+        if (params.hasOwnProperty('utm_medium') && (allowedChars).test(params['utm_medium'])) {
+            finalParams['utm_medium'] = decodeURIComponent(params['utm_medium']);
+        }
+        return Object.getOwnPropertyNames(finalParams).length >= 1 ? finalParams : null;
+    };
+
+    /**
+     * Append an object of URL parameters to a given URL.
+     * @param {String} URL to append parameters to.
+     * @param {Object} data object consisting of one or more parameters.
+     * @returns {String} URL containing updated parameters.
+     */
+    UtmUrl.appendToDownloadURL = function (url, data) {
+        var finalParams;
+        var linkParams;
+
+        if (url.indexOf('?') > 0) {
+            linkParams = window._SearchParams.queryStringToObject(url.split('?')[1]);
+            finalParams = Object.assign(linkParams, data);
+        } else {
+            finalParams = data;
+        }
+
+        return url.split('?')[0] + '?' + window._SearchParams.objectToQueryString(finalParams);
+    };
+
+    /**
+     * If there are valid utm params on the page URL, query the
+     * DOM and update Firefox Account links with the new utm data
+     */
+    UtmUrl.getAttributionData.init = function (){
+        var params = UtmUrl.getAttributionData();
+        var ctaLinks = document.getElementsByClassName('js-fxa-cta-link');
+
+        // If there are no utm params on the page, do nothing.
+        if (!params) {
+            return;
+        }
+
+        // feature detect support for object modification.
+        if (typeof Object.assign !== 'function' || typeof Object.getOwnPropertyNames !== 'function') {
+            return;
+        }
+
+        for (var i = 0; i < ctaLinks.length; i++) {
+            // get the current FxA links.
+            var oldAccountsLink = ctaLinks[i].href;
```

Comment:
I think there's a possible race condition here between this code and [this code](https://github.com/mozilla/bedrock/blob/master/media/js/base/protocol/init-navigation.js#L59-L70). If true, we could end up adding UTM params to an internal link to `/firefox/accounts/`, which would confound our GA data by creating new sessions for those that click the link.

To try and avoid this and make our logic here a bit more robust, it may be worth adding a check to make sure that both `oldAccountsLinks` and `newAccountsLink` both start with `https://accounts.firefox.com/` before updating them. This way, we can be more confident that we're only updating an `href` that points to an external domain. Thoughts?

Reply:
I think you are right, thanks for catching.
Would this be worth separating out into a separate issue? (FxA validation for global nav?)

Reply:
hmm, I'm not certain if we can safely merge this without it causing issues?

Reply:
Or do you mean not applying this logic to the global nav?

Reply:
I mean not applying the logic/`fxa_cta_link` to the nav in this PR and moving it to it's own issue. Maybe that is too granular? Wdyt?

Reply:
I guess it depends how much of the original issue we want to offload for a later task. This is probably a couple of extra lines of JS to add, but we could always skip the navigation button for now if we think that makes sense? My only concern about shipping a partial implementation is that we'll not get time to go back and fix things for a while, or forget to do it.